File: cpp/base/src/main/vsdk/toolkit/render/shaders/CpuTextureSamplingConfig.cpp
#include "CpuTextureSamplingConfig.h"
#include "vsdk/toolkit/common/color/ColorRgb.h"
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "vsdk/toolkit/media/Image.h"
#include "vsdk/toolkit/media/NormalMap.h"

double CpuTextureSamplingConfig::textureOffsetUTexels = -0.5;
double CpuTextureSamplingConfig::textureOffsetVTexels = -0.5;
const double CpuTextureSamplingConfig::NORMAL_OFFSET_U_TEXELS = -0.5;
const double CpuTextureSamplingConfig::NORMAL_OFFSET_V_TEXELS = -0.5;

void CpuTextureSamplingConfig::setTextureOffsetTexels(double uTexels, double vTexels)
{
    textureOffsetUTexels = uTexels;
    textureOffsetVTexels = vTexels;
}

ColorRgb CpuTextureSamplingConfig::sample(Image* texture, double u, double v)
{
    if ( texture == 0 ) return ColorRgb();
    double du = texture->getXSize() > 0 ? textureOffsetUTexels / ((double)texture->getXSize()) : 0.0;
    double dv = texture->getYSize() > 0 ? textureOffsetVTexels / ((double)texture->getYSize()) : 0.0;
    ColorRgb* c = texture->getColorRgbBiLinear(u + du, v + dv);
    ColorRgb out = c != 0 ? *c : ColorRgb();
    delete c;
    return out;
}

Vector3Dd CpuTextureSamplingConfig::sampleNormal(NormalMap* normalMap, double u, double v)
{
    if ( normalMap == 0 ) return Vector3Dd();
    double du = normalMap->getXSize() > 0 ? NORMAL_OFFSET_U_TEXELS / ((double)normalMap->getXSize()) : 0.0;
    double dv = normalMap->getYSize() > 0 ? NORMAL_OFFSET_V_TEXELS / ((double)normalMap->getYSize()) : 0.0;
    Vector3Dd* n = normalMap->getNormalBiLinear(u + du, v + dv);
    Vector3Dd out = n != 0 ? *n : Vector3Dd();
    delete n;
    return out;
}

File: typescript/base/src/main/vsdk/toolkit/render/shaders/CpuTextureSamplingConfig.ts
import type { ColorRgb } from "../../common/color/ColorRgb.js";
import type { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import type { Image } from "../../media/Image.js";
import type { NormalMap } from "../../media/NormalMap.js";

export class CpuTextureSamplingConfig {
    private static textureOffsetUTexels = -0.5;
    private static textureOffsetVTexels = -0.5;
    private static readonly NORMAL_OFFSET_U_TEXELS = -0.5;
    private static readonly NORMAL_OFFSET_V_TEXELS = -0.5;

    private constructor() {}

    public static setTextureOffsetTexels(uTexels: number, vTexels: number): void {
        CpuTextureSamplingConfig.textureOffsetUTexels = uTexels;
        CpuTextureSamplingConfig.textureOffsetVTexels = vTexels;
    }

    public static sample(texture: Image | null, u: number, v: number): ColorRgb | null {
        if (texture === null) {
            return null;
        }
        let du = 0.0;
        let dv = 0.0;
        const width: number = texture.getXSize();
        const height: number = texture.getYSize();
        if (width > 0) {
            du = CpuTextureSamplingConfig.textureOffsetUTexels / width;
        }
        if (height > 0) {
            dv = CpuTextureSamplingConfig.textureOffsetVTexels / height;
        }
        return texture.getColorRgbBiLinear(u + du, v + dv);
    }

    public static sampleNormal(normalMap: NormalMap | null, u: number, v: number): Vector3Dd | null {
        if (normalMap === null) {
            return null;
        }
        let du = 0.0;
        let dv = 0.0;
        const width: number = normalMap.getXSize();
        const height: number = normalMap.getYSize();
        if (width > 0) {
            du = CpuTextureSamplingConfig.NORMAL_OFFSET_U_TEXELS / width;
        }
        if (height > 0) {
            dv = CpuTextureSamplingConfig.NORMAL_OFFSET_V_TEXELS / height;
        }
        return normalMap.getNormal(u + du, v + dv);
    }
}
